### PR TITLE
Modularise keeper rewards

### PR DESCRIPTION
**Pull request by CalabashSquash**
# Motivation

Keeper rewards can be upgraded in the future.

### CODE REVIEW

File: contracts/implementation/KeeperRewards.sol [hunk 1]
```diff
@@ -0,0 +1,165 @@
+//SPDX-License-Identifier: CC-BY-NC-ND-4.0
+pragma solidity 0.8.7;
+
+import "../interfaces/ILeveragedPool.sol";
+import "../interfaces/IKeeperRewards.sol";
+import "../interfaces/IOracleWrapper.sol";
+
+import "../libraries/PoolSwapLibrary.sol";
+
+/// @title The pool contract itself
```
Comment (chrdevmar) [inline contracts/implementation/KeeperRewards.sol]: this comment looks wrong

File: contracts/implementation/KeeperRewards.sol [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+     * @param _gasSpent Number of gas units spent
+     * @return Keeper's gas compensation
+     * @dev Adds a constant to `_gasSpent` when calculating actual gas usage
+     */
+    function keeperGas(
+        uint256 _gasPrice,
+        uint256 _gasSpent,
+        uint256 _settlementTokenPrice
+    ) public pure returns (uint256) {
+        if (_settlementTokenPrice <= 0) {
```
Comment (chrdevmar) [inline contracts/implementation/KeeperRewards.sol]: do you know if `==` is cheaper than `<=`? if so could we use that? its unsigned anyway

File: contracts/implementation/PoolKeeper.sol [hunk 3]
```diff
@@ -137,18 +131,29 @@ contract PoolKeeper is IPoolKeeper, Ownable {
...
         } catch Error(string memory reason) {
             // If poolUpkeep fails for any other reason, emit event
             emit PoolUpkeepError(_pool, reason);
         }
+
+        uint256 gasSpent = startGas - gasleft();
+        uint256 reward;
+        // Emit events depending on whether or not the reward was actually paid
+        if (
+            IKeeperRewards(keeperRewards).payKeeper(
```
Comment (dospore) [inline contracts/implementation/PoolKeeper.sol:144]: should this get the keeper from the Pool contract instead?? Assuming you could drop the setKeeper from this contract as well?

File: contracts/implementation/KeeperRewards.sol [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+            emit PoolUpkeepError(_pool, reason);
+        }
+        int256 settlementTokenPrice = IOracleWrapper(ILeveragedPool(_pool).settlementEthOracle()).getPrice();
+
+        uint256 reward = keeperReward(
+            _gasPrice,
+            _gasSpent,
+            _savedPreviousUpdatedTimestamp,
+            _updateInterval,
+            uint256(settlementTokenPrice)
```
Comment (chrdevmar) [inline contracts/implementation/KeeperRewards.sol:61]: do you know what the behaviour is when casting a negative `int256` to a `uint256`?

File: contracts/implementation/PoolKeeper.sol (hunk 3 above)
Comment (CalabashSquash) [reply contracts/implementation/PoolKeeper.sol:144]: What do you mean by "get the keeper from the Pool contract"?
As in, get `msg.sender` from inside a `LeveragedPool` context?

Comment (dospore) [reply contracts/implementation/PoolKeeper.sol:144]: no, here its initialising the keeper with keeperRewards from PoolKeeper.sol, wouldnt you want `ILeveragedPool(_pool).keeperRewards`

Comment (dospore) [reply contracts/implementation/PoolKeeper.sol:144]: or you could drop the `setKeeperRewards` and `keeperRewards` LeveragedPool.sol, and use `keeper.keeperRewards()` in `onlyKeeperRewards` the double up is confusing me, but maybe its cheaper

Comment (dospore) [reply contracts/implementation/PoolKeeper.sol:144]: but maybe there is no way around it

Comment (CalabashSquash) [reply contracts/implementation/PoolKeeper.sol:144]: Oooh you just made me remember a dream last night where I removed `LeveragedPool::keeperRewards` and just called `PoolKeeper.keeperRewards()` in `onlyKeeperRewards`. Similar to what you're saying but kind of flipped the other way. This way we don't have to update every pool when we change `keeperRewards`.

**Review by SaadmanM — APPROVED** [review]

File: contracts/implementation/KeeperRewards.sol [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+     * @notice Tip a keeper will receive for successfully updating the specified pool
+     * @param _savedPreviousUpdatedTimestamp Last timestamp when the pool's price execution happened
+     * @param _poolInterval Pool interval of the given pool
+     * @return Percent of the `keeperGas` cost to add to payment, as a percent
+     */
+    function keeperTip(uint256 _savedPreviousUpdatedTimestamp, uint256 _poolInterval) public view returns (uint256) {
+        /* the number of blocks that have elapsed since the given pool's updateInterval passed */
+        uint256 elapsedBlocksNumerator = (block.timestamp - (_savedPreviousUpdatedTimestamp + _poolInterval));
+
+        uint256 keeperTipAmount = BASE_TIP + (TIP_DELTA_PER_BLOCK * elapsedBlocksNumerator) / BLOCK_TIME;
```
Comment (chrdevmar) [inline contracts/implementation/KeeperRewards.sol:156]: interested as to why we are using L1 block times here (or is it just coincidence?). 

we could make the number less "magic" by doing something like:

```
uint256 secondsOverdue = (block.timestamp - (_savedPreviousUpdatedTimestamp + _poolInterval));
uint256 percentOverdue = (secondsOverdue * 100) / _poolInterval;

uint256 keeperTipAmount = BASE_TIP + (TIP_DELTA_PER_PERCENT_OVERDUE * percentOverdue);
```

File: contracts/implementation/KeeperRewards.sol (hunk 1 above)
Comment (CalabashSquash) [reply contracts/implementation/KeeperRewards.sol]: Fixed

File: contracts/implementation/KeeperRewards.sol (hunk 2 above)
Comment (CalabashSquash) [reply contracts/implementation/KeeperRewards.sol]: Fixed

File: contracts/implementation/KeeperRewards.sol (hunk 4 above)
Comment (CalabashSquash) [reply contracts/implementation/KeeperRewards.sol:61]: reverts

File: contracts/implementation/KeeperRewards.sol (hunk 5 above)
Comment (CalabashSquash) [reply contracts/implementation/KeeperRewards.sol:156]: Good point. out of scope

Comment (chrdevmar) [reply contracts/implementation/KeeperRewards.sol:156]: yep agree, good thing we can change this now

**Review by chrdevmar — APPROVED** [review]

**Review by sporejack — APPROVED** [review]

**Review by SaadmanM — APPROVED** [review]

**Pull request merged**